src/Front/Scope.hs

```diff
@@ -0,0 +1,311 @@
+data ScopeCtx = ScopeCtx
+  { dataMap :: M.Map Identifier DataInfo
+  , typeMap :: M.Map Identifier TypInfo
+  }
+
+-- | Scoping monad.
+newtype ScopeFn a = ScopeFn (ReaderT ScopeCtx Pass a)
+  deriving Functor                      via (ReaderT ScopeCtx Pass)
+  deriving Applicative                  via (ReaderT ScopeCtx Pass)
+  deriving Monad                        via (ReaderT ScopeCtx Pass)
+  deriving MonadFail                    via (ReaderT ScopeCtx Pass)
+  deriving (MonadError Error)           via (ReaderT ScopeCtx Pass)
+  deriving (MonadWriter [Warning])      via (ReaderT ScopeCtx Pass)
+  deriving (MonadReader ScopeCtx)       via (ReaderT ScopeCtx Pass)
+
+-- | Run a ScopeFn computation.
+runScopeFn :: ScopeFn a -> Pass a
+runScopeFn (ScopeFn m) =
+  runReaderT m ScopeCtx { dataMap = builtinData, typeMap = builtinTypes }
+
+-- | Add a list of data identifiers to the scope.
+withDataScope :: [(Identifier, DataInfo)] -> ScopeFn a -> ScopeFn a
+withDataScope is = do
+  local $ \ctx -> ctx { dataMap = foldr (uncurry M.insert) (dataMap ctx) is }
+
+-- | Check that an 'Identifier' is not an empty string.
+ensureNonempty :: Identifier -> ScopeFn ()
+ensureNonempty i =
+  when (null $ ident i) $ throwError $ UnexpectedError "Empty identifier"
+
+-- | Check that a set of bindings does not define overlapping 'Identifier'.
+ensureUnique :: [Identifier] -> ScopeFn ()
+ensureUnique ids = do
+  forM_ (group $ sort ids) $ \case
+    []    -> throwError $ UnexpectedError "unique should not be empty"
+    [ _ ] -> return ()
+    i : _ -> throwError $ ScopeError $ "Defined more than once: " <> showId i
+
+-- | Check that a constructor 'Identifier' has the right naming convention.
+ensureCons :: Identifier -> ScopeFn ()
+ensureCons i = do
+  ensureNonempty i
+  unless (isCons i) $ throwError nameErr
+ where
+  nameErr =
+    NameError
+      $  showId i
+      <> " should begin with upper case or begin and end with ':'"
+
+-- | Check that a variable 'Identifier' has the right naming convention.
+ensureVar :: Identifier -> ScopeFn ()
+ensureVar i = do
+  ensureNonempty i
+  unless (isVar i) $ throwError nameErr
+ where
+  nameErr =
+    NameError
+      $  showId i
+      <> " should begin with upper case or begin and end with ':'"
+
+{- | Validate a declaration of a data 'Identifier'.
+
+This includes binding to the left of the @=@ in a let-binding or pattern match,
+and in the argument of lambdas. Does not include type definitions, where data
+constructors are defined.
+-}
+dataDecl :: Identifier -> ScopeFn ()
+dataDecl i = do
+  info <- asks $ M.lookup i . dataMap
+
+  -- A data constructor cannot be defined inside of a pattern.
+  when (isCons i && not (inScope info)) $ do
+    throwError $ ScopeError $ "Data constructor is out of scope: " <> showId i
+
+  -- A data variable can usually be shadowed, but we want warn about it.
+  when (isVar i && inScope info) $ if canShadow info
+    then warn $ NameWarning $ "shadowing variable: " <> showId i
```

Comment:
Just added support for warnings --- now we can just `warn $ KindOfWarning "warn message"` from anywhere within the Pass monad to throw a warning:

```
$ stack exec sslc -- - >/dev/null <<-EOF
f (x: Int) -> Int =
  let (x: Int) = 3
  x
EOF
[NameWarning (ErrorMsg "shadowing variable: 'x'")]
```

Right now it just dumps the default Show instance, which is ugly, but now we can easily fine-tune the CLI reporting for warnings and errors by creating a Pretty instance and using that